Exercise 5

Marcelloloco opened the pull request:
Rough version, still some fixes necessary

Marcelloloco committed:
[ADD] npm packages for neo4j driver and neo4j-graphql-js

Marcelloloco committed:
[ADD] Implementation of neo4j schema and neo4j-driver. Cypher statements added
for hooking up neo4j database.

Marcelloloco committed:
[ADD] Funcionality for neo4j for every mutation and adjustments for the tests and Github Actions

Marcelloloco committed:
[FIX] Adjustment to before each and after all functions for keeping the database clean.

Marcelloloco committed:
[FIX] lint fixes

Marcelloloco committed:
[FIX] Cleaning for tests

Marcelloloco committed:
[FIX] Refactoring of datasource and fix of tests and linter

All test are running now in a persistent way with the Neo4j database.

Marcelloloco committed:
[FIX] Detach docker run to not block further execution

Marcelloloco committed:
[FIX] Delete function Neo4j implementation

Marcelloloco commented:
Review was done for the guys from @Systems-Development-and-Frameworks/thecodeninjas .

PR: https://github.com/Systems-Development-and-Frameworks/TheCodeNinjas/pull/8

Marcelloloco committed:
[ADD] Nested Queries for Neo4j

Queries are now processed with delegateToSchema and permission for sub-queries are set.

RespectableRuessel commented on backend/.env.test:
```diff
@@ -1 +1,3 @@
-JWT_SECRET=GitTestKeyTZGHbb_iuidsuiwew
\ No newline at end of file
+JWT_SECRET=GitTestKeyTZGHbb_iuidsuiwew
+NEO4J_USERNAME=neo4j
+NEO4J_PASSWORD=neo4j
```
Maybe add a NEO4J_DATABASE key if the database is not the default database.

TheKilljoy replied:
:thumbsup: Using separate environment file, and no sensitive keys committed.

RespectableRuessel commented on backend/src/datasources/userDatasource.js:
```diff
@@ -45,6 +49,20 @@ export class UserDataSource extends DataSource {
   }
 
   async getAll () {
+    this.users = []
+    const session = driver.session()
+    const txc = session.beginTransaction()
+
+    try {
+      const result = await txc.run('MATCH(u:User) RETURN u')
+      await txc.commit()
+      result.records.forEach(r => this.users.push(r.get(0).properties))
+    } catch (error) {
+      console.log(error)
+    } finally {
+      await session.close()
+    }
+
```
This is dangerous 🔥  Your method prefix is "get" but has a side effect. This is dangerous because you're overwriting the `this.user` variable thus manipulating the state of the object. Also every other method need to call `getAll` before accessing `this.user` because otherwise it is not guaranteed to hold the latest data.
Also this behavior is not documented so if you miss calling the method before a `this.user` call your state can be broken.

RespectableRuessel commented on backend/src/datasources/userDatasource.js:
```diff
@@ -56,6 +74,8 @@ export class UserDataSource extends DataSource {
 
```
`removeUser` misses the ` await this.getAll()` call. Also the removal has no impact on the data in your database. It looks like no user is actually removed from the database.

RespectableRuessel reviewed (changes requested):
⭐ For choosing a scenario and writing installation instructions in README.md.

⭐ For explaining WHY you have chosen the scenario in the README.md.

⭐ For not committing sensitive secrets (e.g. API keys) unencrypted to the repository.

🟥 Data created in mutations are persisted and your queries and mutations still work as expected.

See comments on `getAll` and `removeUser`.

⭐ ⭐ Your software tests are free of side effects.

🟥 Your database is never left in an invalid state.

See email verification.

⭐ You deny the client to call mutations and queries of your subschema directly.

⭐ For requesting a review and reviewing another team's PR.

TheKilljoy commented on backend/README.md:
```diff
@@ -1,4 +1,6 @@
 # backend
+In this part of the project we decided to work with a Neo4j database, because the graph based
+approach seemed very interesting and we wanted to play around with it. 
```
👍 Reason given for why Neo4j was chosen.

medizinmensch replied:
Why is it interesting? Why isn't the other one interesting? Why did you want to play around with neo4j and not GraphCMS?
Next time, please take elaborate on objectives like this.
:star:

TheKilljoy commented on backend/README.md:
````diff
@@ -7,6 +9,39 @@ cd backend
 ```
 npm install
 ```
+The backend of this application is hooked up to a Neo4j Database. In order to work with it, you can
+install Docker on your computer and run a Neo4j container.
+
+For installing Docker on an Ubuntu machine you could use:
+```
+sudo apt-get update
+sudo apt-get install docker-ce docker-ce-cli containerd.io
+```
+
+For installation on an Arch Linux based system use:
+```
+sudo pacman -Syu
+sudo pacman -S docker
+```
+
+For installation on Windows...
+
+...but who uses that anyway ;)
+
+Once Docker is up and running all you need to do is start the Neo4j container.
+You can pull it simply via:
+```
+docker run  --publish=7474:7474 --publish=7687:7687 --env=NEO4J_AUTH=none neo4j:latest
+```
+
+At last you need to provide the credentials for Neo4j in your .env file, which could
+look like this:
+```
+NEO4J_USERNAME=neo4j
+NEO4J_PASSWORD=neo4j
+```
+
+All your GraphQL queries and mutations will be directed to this Database.
````
👍 Steps for installation of docker in different OS, deployment of Neo4j via docker and configuarion of the .env file are present.

RespectableRuessel commented on backend/src/datasources/postDatasource.js:
```diff
@@ -48,6 +64,20 @@ export class PostDataSource extends DataSource {
   }
 
   async getAll () {
+    this.posts = []
+    const session = driver.session()
+    const txc = session.beginTransaction()
+    try {
+      const result = await txc.run('MATCH(p:Post) RETURN p')
+      await txc.commit()
+      result.records.forEach(r => { r.get(0).properties.votes = r.get(0).properties.votes.toNumber() })
+      result.records.forEach(r => this.posts.push(r.get(0).properties))
+    } catch (error) {
+      console.log(error)
+    } finally {
+      await session.close()
+    }
+
```
🔥 Same problem as in `userDatasource`.

TheKilljoy commented on backend/src/datasources/userDatasource.js:
```diff
@@ -56,6 +74,8 @@ export class UserDataSource extends DataSource {
     const saltRounds = 10
+    await this.getAll()
+
     if (this.users.some((user) => user.email === args.email)) {
       throw new UserInputError('This email already exists')
     }
```
👎 E-Mails should at least be checked for an existing '@' symbol inside the string. Also empty strings are still valid.

TheKilljoy reviewed (changes requested):
- Check E-Mail for correctness

Marcelloloco committed:
[ADD] Schema Delegation and refactoring

Schema delegation is now used where it is possible, generated cypher statements are
blocked by graphql shield and test cases are adjusted to more persistent approach therefore
the mock test have been removed.

Marcelloloco committed:
[REMOVED] Mock tests

Marcelloloco committed:
[Fix] Cleanup of unused code

Remove the getAll functions which are not used anymore

Marcelloloco commented:
@RespectableRuessel @TheKilljoy thanks for the Feedback! Schema delegation and custom calls for certain users/posts are now used instead of the old getAll() Method. Tests have also been adjusted and mocks removed because it is all persistent now.

medizinmensch commented on backend/README.md:
````diff
@@ -7,6 +9,39 @@ cd backend
+docker run  --publish=7474:7474 --publish=7687:7687 --env=NEO4J_AUTH=none neo4j:latest
+```
+
+At last you need to provide the credentials for Neo4j in your .env file, which could
````
I _really_ like your detailed Readme on how to setup your backend. Here is a critical typo though. Maybe you should elaborate that `npm run test` is using the `.env.test` file whereas `dev` is using the `.env.development` file.
```suggestion
At last you need to provide the credentials for Neo4j in your .env.test file, which could
```

medizinmensch commented on backend/.env.test:
```diff
@@ -1 +1,3 @@
-JWT_SECRET=GitTestKeyTZGHbb_iuidsuiwew
\ No newline at end of file
+JWT_SECRET=GitTestKeyTZGHbb_iuidsuiwew
```
This looks like an actual secret? I hope/assume it's just for testing/development. 
:star: For not committing   sensitive secrets (e.g. API keys) unencrypted to the repository.

medizinmensch commented on backend/src/datasources/postDatasource.js:
```diff
@@ -62,13 +121,34 @@ export class PostDataSource extends DataSource {
+      await session.close()
+    }
+
     author.posts.push(newPost)
     this.posts.push(newPost)
     return newPost
   }
```
It would be better to ask the database about the post/author. If the database - for some reason - did not save your transaction properly (or changed it), the client wouldn't know as your pushing the post to `this.post` as well as `author.posts`.
You should not have a state in this class at all. 
`this.posts` should get removed from the class, it's constructor as well as everywhere else. 
I see that you already removed it most of the places. Maybe you just forgot the last 3 occurrences?

medizinmensch reviewed (approved):
Hi Team!

First of all, a thanks to @RespectableRuessel for finding the issues with the persistence and the database state. 

Please read my comments below - there are still some things you should be aware of.
From a points perspective:


:star: For choosing a scenario   and writing installation instructions in README.md.
:star:   For explaining WHY you have chosen the scenario in the README.md.
:star: For not committing   sensitive secrets (e.g. API keys) unencrypted to the repository.
:star:   :star: Data created in mutations are persisted and your queries and mutations   still work as expected.
:star: :star: Your software   tests are free of side effects.
:star:   Your database is never left in an invalid state.
:star: You deny the client to   call mutations and queries of your subschema directly.
:star:   For requesting a review and reviewing another team's PR.

That makes **10/10** :star:

Congrats!

![Congrats](https://media.giphy.com/media/DKnMqdm9i980E/giphy.gif)
